GPU/kmeans_util_sa.cu

```diff
@@ -0,0 +1,521 @@
+
+    while(curr_k < k-1) {
+        max = -1;
+        max_i = -1;
+        for(i=0; i<n; i++){
+            DPRINTF("New squared_distance: %f and old min squared_distance: %f", squared_distance(ps[i], centers[curr_k], dim), distances_from_centers[i]);
+            temp_distances[i] = MIN(squared_distance(ps[i], centers[curr_k], dim), distances_from_centers[i]);  
+            if(temp_distances[i] > max){
+                max = temp_distances[i];
+                max_i = i;
+            }
+        }
+ 
+        memcpy(distances_from_centers, temp_distances, n * sizeof(double));
+        memcpy(centers[++curr_k], ps[max_i], dim * sizeof(double));
+    }
+    
+    free(temp_distances);
+    free(distances_from_centers);
+    return centers;
+}
+
+__global__
+void find_cluster_on_gpu(const double *dev_points, const double *dev_centers, 
+                         const int n, const int k, const int dim, 
+                         double *result_clusters) {
+
+    double min, dist;
+    int cluster_it_belongs;
+    const unsigned int index = get_global_tid();
+    const unsigned int thread_id = threadIdx.x;
+    extern __shared__ double local_centers[];
+
+    const unsigned int start = index;
+    const unsigned int end = start + 1;
+
+    // WARNING: Mporei na dhmiourgithei provlhma an ta threads sto block einai ligotera apo to k*dim
+    if(thread_id < k*dim){
+        local_centers[thread_id] = dev_centers[thread_id];
+    }
+    __syncthreads();
+
+    if (index < n){
+        for (int i = start; i < end; i++){
+            min = DBL_MAX;
+            for (int j = 0; j < k; j++){
+                result_clusters[j*n + i] = 0.0;
+                dist = squared_distance_on_gpu(&dev_points[i], &local_centers[j], n, k, dim);
+
+                if (min > dist){
+                    min = dist;
+                    cluster_it_belongs = j;
+                }
+                // cluster_it_belongs = j*(min > dist) + cluster_it_belongs*(min <= dist);
+                // min = min*(min <= dist) + dist*(min > dist);
+                // cluster_it_belongs = cluster_it_belongs ^ ((j ^ cluster_it_belongs) & -(min < dist));
+                // min = dist ^ ((min ^ dist) & -(min < dist));
+            }
+            // Only 1 in the cluster it belongs and everything else 0
+            // result_clusters[cluster_it_belongs*n + i] = 1.0;
+            result_clusters[cluster_it_belongs*n + i] = 1.0;
+            // for (int j = 0; j < k; j++){
+            //     printf("result_clusters[%d][%d] = %lf --> line[%d]\n", j, i, result_clusters[j*n + i], i+2);
+            // }
+        }
+    }
+}
+
+__global__ void setup_RNG_states(curandState *devStates,  unsigned long seed){
+    const unsigned int index = get_global_tid();
+    //Init curand for each thread
+    curand_init(seed, index, 0, &devStates[index]);
+}
+
+__device__ double calc_SA_prob(double de,  double dmin,  curandState* state){
+    double prop = exp(-abs(de-dmin)/sa_temp);
```

Comment:
Is there any better way of computing this probablity and also choosing yes or no without the if?? (bitwise ops??)